packages/wallet/src/redux/indirect-funding/player-a/reducer.ts

```diff
@@ -18,19 +39,365 @@ export function playerAReducer(
 
   switch (state.indirectFunding.type) {
     case states.WAIT_FOR_APPROVAL:
+      return waitForApprovalReducer(state, action);
     case states.WAIT_FOR_PRE_FUND_SETUP_1:
-    // pass the commitment to the channel state reducer
-    // update channel state based on outcome
-    // if ready to fund:
-    //   initiate direct funding
+      return waitForPreFundSetup1Reducer(state, action);
     case states.WAIT_FOR_DIRECT_FUNDING:
-    // progress direct funding
-    // if direct funding is finished:
-    //   send and/or wait for post-fund setup
+      return waitForDirectFunding(state, action);
     case states.WAIT_FOR_POST_FUND_SETUP_1:
+      return waitForPostFundSetup1(state, action);
     case states.WAIT_FOR_LEDGER_UPDATE_1:
-      return state;
+      return waitForLedgerUpdateReducer(state, action);
     default:
       return unreachable(state.indirectFunding);
   }
 }
+
+const waitForLedgerUpdateReducer = (
+  state: walletStates.Initialized,
+  action: actions.indirectFunding.Action,
+): walletStates.Initialized => {
+  switch (action.type) {
+    case actions.COMMITMENT_RECEIVED:
+      const indirectFundingState = selectors.getIndirectFundingState(
+        state,
+      ) as states.WaitForLedgerUpdate1;
+
+      // Update ledger state
+      let newState = receiveLedgerCommitment(state, action.commitment, action.signature);
+      if (
+        ledgerChannelFundsAppChannel(
+          newState,
+          indirectFundingState.channelId,
+          indirectFundingState.ledgerId,
+        )
+      ) {
+        newState = confirmFundingForAppChannel(newState, indirectFundingState.channelId);
+      }
+      return newState;
+    default:
+      return state;
+  }
+};
+
+const waitForPostFundSetup1 = (
+  state: walletStates.Initialized,
+  action: actions.indirectFunding.Action,
+): walletStates.Initialized => {
+  switch (action.type) {
+    case actions.COMMITMENT_RECEIVED:
+      const indirectFundingState = selectors.getIndirectFundingState(
+        state,
+      ) as states.WaitForPostFundSetup1;
+
+      let newState = receiveLedgerCommitment(state, action.commitment, action.signature);
+
+      if (ledgerChannelIsWaitingForUpdate(newState, indirectFundingState.ledgerId)) {
+        newState = createAndSendUpdateCommitment(
+          newState,
+          indirectFundingState.channelId,
+          indirectFundingState.ledgerId,
+        );
+        newState.indirectFunding = states.waitForLedgerUpdate1(indirectFundingState);
+      }
+
+      return newState;
+
+    default:
+      return state;
+  }
+};
+const waitForDirectFunding = (
+  state: walletStates.Initialized,
+  action: actions.indirectFunding.Action,
+): walletStates.Initialized => {
+  const indirectFundingState = selectors.getIndirectFundingState(
+    state,
+  ) as states.WaitForDirectFunding;
+  if (!isFundingAction(action) || action.channelId !== indirectFundingState.ledgerId) {
```

Comment:
hmm, shouldn't the action's `channelId` be the channel id that the procedure is related to? Which, if it's indirect funding, is the app-channel id?

Reply:
I think you're right. I've updated this.

However this also brings to light that we have to figure out how to attach the correct channelID to the `FUNDING_RECEIVED_EVENT`.